linkerd/dns/src/lib.rs

```diff
@@ -98,11 +99,25 @@ impl Resolver {
                         .unwrap_or_else(|| std::time::Instant::now() + Self::DEFAULT_TTL);
                     Ok((vec![], time::delay_until(Instant::from_std(expiry))))
                 }
+                ResolveErrorKind::Proto(pe) if self.is_nx_domain(&pe) => Ok((
+                    vec![],
+                    time::delay_until(Instant::from_std(
+                        std::time::Instant::now() + Self::DEFAULT_TTL,
+                    )),
+                )),
+
                 _ => Err(e),
             },
         }
     }
 
+    fn is_nx_domain(&self, pe: &error::ProtoError) -> bool {
+        return match pe.kind() {
+            error::ProtoErrorKind::Message(msg) => *msg == "Nameserver responded with NXDomain",
+            _ => false,
+        };
```

Comment:
nit: the `return` is not necessary, this can just be
```suggestion
        match pe.kind() {
            error::ProtoErrorKind::Message(msg) => *msg == "Nameserver responded with NXDomain",
            _ => false,
        }
```